src/tests/unittests/d3d12/CopySplit.cpp

```diff
@@ -0,0 +1,375 @@
+        }
+        return os;
+    }
+
+    // Define base texture sizes and offsets to test with: some aligned, some unaligned
+    constexpr TextureSpec kBaseTextureSpecs[] = {
+        { 0, 0, 0, 1, 1, 1, 4 },
+        { 31, 16, 0, 1, 1, 1, 4 },
+        { 64, 16, 0, 1, 1, 1, 4 },
+
+        { 0, 0, 0, 1024, 1024, 1, 4 },
+        { 256, 512, 0, 1024, 1024, 1, 4 },
+        { 64, 48, 0, 1024, 1024, 1, 4 },
+
+        { 0, 0, 0, 257, 31, 1, 4 },
+        { 0, 0, 0, 17, 93, 1, 4 },
+        { 59, 13, 0, 257, 31, 1, 4 },
+        { 17, 73, 0, 17, 93, 1, 4 },
+    };
+
+    // Define base buffer sizes to work with: some offsets aligned, some unaligned. rowPitch is the minimum required
+    std::array<BufferSpec, 10> BaseBufferSpecs(const TextureSpec& textureSpec) {
+        uint32_t rowPitch = Align(textureSpec.texelSize * textureSpec.width, kTextureRowPitchAlignment);
+
+        auto alignNonPow2 = [](uint32_t value, uint32_t size) -> uint32_t {
+            return value == 0 ? 0 : ((value - 1) / size + 1) * size;
+        };
+
+        return {
+             alignNonPow2(0, textureSpec.texelSize), rowPitch,
+             alignNonPow2(512, textureSpec.texelSize), rowPitch,
+             alignNonPow2(1024, textureSpec.texelSize), rowPitch,
+
+             alignNonPow2(32, textureSpec.texelSize), rowPitch,
+             alignNonPow2(64, textureSpec.texelSize), rowPitch,
+            
+             alignNonPow2(31, textureSpec.texelSize), rowPitch,
+             alignNonPow2(257, textureSpec.texelSize), rowPitch,
+             alignNonPow2(511, textureSpec.texelSize), rowPitch,
+             alignNonPow2(513, textureSpec.texelSize), rowPitch,
+             alignNonPow2(1023, textureSpec.texelSize), rowPitch,
+        };
+    }
+
+    // Define a list of values to set properties in the spec structs
+    constexpr uint32_t kCheckValues[] = {
+        1, 2, 3, 4, 5, 6, 7, 8,                // small values
+        16, 32, 64, 128, 256, 512, 1024, 2048, // powers of 2
+        15, 31, 63, 127, 257, 511, 1023, 2047, // misalignments
+        17, 33, 65, 129, 257, 513, 1025, 2049
+    };
+
+}
+
+class CopySplitTest : public testing::Test {
+    protected:
+        CopySplit DoTest(const TextureSpec& textureSpec, const BufferSpec& bufferSpec) {
+            CopySplit copySplit = ComputeSplit(textureSpec.x, textureSpec.y, textureSpec.z, textureSpec.width, textureSpec.height, textureSpec.depth, textureSpec.texelSize, bufferSpec.offset, bufferSpec.rowPitch);
+            ValidateCopySplit(textureSpec, bufferSpec, copySplit);
+            return copySplit;
+        }
+};
+
+TEST_F(CopySplitTest, General) {
+    for (TextureSpec textureSpec : kBaseTextureSpecs) {
+        for (BufferSpec bufferSpec : BaseBufferSpecs(textureSpec)) {
+
+            CopySplit copySplit = DoTest(textureSpec, bufferSpec);
+            if (HasFatalFailure()) {
```

Comment:
I want to print which split failed, and then if so exit the test. If I don't exit immediately, it's going to print for every failure after the first failure. If I factor it out, then the FAIL() macro won't exit the test because the result of the FAIL will be wrapped in another function